// proteus-js-frames/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _ProteusTypes = require('./ProteusTypes');

Object.keys(_ProteusTypes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _ProteusTypes[key];
    }
  });
});

var _ProteusFrame = require('./ProteusFrame');

Object.defineProperty(exports, 'FrameTypes', {
  enumerable: true,
  get: function get() {
    return _ProteusFrame.FrameTypes;
  }
});
Object.defineProperty(exports, 'getFrameTypeName', {
  enumerable: true,
  get: function get() {
    return _ProteusFrame.getFrameTypeName;
  }
});

var _ProteusBinaryFraming = require('./ProteusBinaryFraming');

Object.defineProperty(exports, 'encodeFrame', {
  enumerable: true,
  get: function get() {
    return _ProteusBinaryFraming.encodeFrame;
  }
});
Object.defineProperty(exports, 'decodeFrame', {
  enumerable: true,
  get: function get() {
    return _ProteusBinaryFraming.decodeFrame;
  }
});

var _ProteusMetadata = require('./ProteusMetadata');

Object.defineProperty(exports, 'encodeProteusMetadata', {
  enumerable: true,
  get: function get() {
    return _ProteusMetadata.encodeProteusMetadata;
  }
});
Object.defineProperty(exports, 'getVersion', {
  enumerable: true,
  get: function get() {
    return _ProteusMetadata.getVersion;
  }
});
Object.defineProperty(exports, 'getService', {
  enumerable: true,
  get: function get() {
    return _ProteusMetadata.getService;
  }
});
Object.defineProperty(exports, 'getMethod', {
  enumerable: true,
  get: function get() {
    return _ProteusMetadata.getMethod;
  }
});
Object.defineProperty(exports, 'getMetadata', {
  enumerable: true,
  get: function get() {
    return _ProteusMetadata.getMetadata;
  }
});
Object.defineProperty(exports, 'getTracing', {
  enumerable: true,
  get: function get() {
    return _ProteusMetadata.getTracing;
  }
});
