// proteus-js-client/dist/index.js
/**
 * Copyright (c) 2017-present, Netifi Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * 
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Empty = exports.Event = exports.Destination = exports.Group = exports.Broker = exports.BrokerManagementServiceServer = exports.BrokerManagementServiceClient = exports.BrokerInfoServiceServer = exports.BrokerInfoServiceClient = exports.toObservable = exports.Proteus = undefined;

var _Proteus = require('./Proteus');

var _Proteus2 = _interopRequireDefault(_Proteus);

var _broker_info_pb = require('./proteus/broker_info_pb');

var _empty_pb = require('google-protobuf/google/protobuf/empty_pb.js');

var _broker_info_proteus_pb = require('./proteus/broker_info_proteus_pb');

var _broker_mgmt_proteus_pb = require('./proteus/broker_mgmt_proteus_pb');

var _FlowableAdapter = require('./rx/FlowableAdapter');

var _FlowableAdapter2 = _interopRequireDefault(_FlowableAdapter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * The public API of the `client` package.
 */
exports.Proteus = _Proteus2.default;
exports.toObservable = _FlowableAdapter2.default;
exports.BrokerInfoServiceClient = _broker_info_proteus_pb.BrokerInfoServiceClient;
exports.BrokerInfoServiceServer = _broker_info_proteus_pb.BrokerInfoServiceServer;
exports.BrokerManagementServiceClient = _broker_mgmt_proteus_pb.BrokerManagementServiceClient;
exports.BrokerManagementServiceServer = _broker_mgmt_proteus_pb.BrokerManagementServiceServer;
exports.Broker = _broker_info_pb.Broker;
exports.Group = _broker_info_pb.Group;
exports.Destination = _broker_info_pb.Destination;
exports.Event = _broker_info_pb.Event;
exports.Empty = _empty_pb.Empty;
