// proteus-js-client/dist/tracing/ProteusTracing.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapToBuffer = mapToBuffer;
exports.bufferToMap = bufferToMap;
exports.trace = trace;
exports.traceAsChild = traceAsChild;
exports.traceSingle = traceSingle;
exports.traceSingleAsChild = traceSingleAsChild;

var _rsocketCore = require('rsocket-core');

var _rsocketFlowable = require('rsocket-flowable');

var _SpanSubscriber = require('./SpanSubscriber');

var _opentracing = require('opentracing');

function mapToBuffer(map) {
  if (!map || Object.keys(map).length <= 0) {
    return (0, _rsocketCore.createBuffer)(0);
  }

  var aggregatedTags = Object.keys(map).reduce(function (aggregate, key) {
    var val = map[key];
    var keyLen = _rsocketCore.UTF8Encoder.byteLength(key);
    var keyBuf = (0, _rsocketCore.createBuffer)(keyLen);
    _rsocketCore.UTF8Encoder.encode(key, keyBuf, 0, keyLen);

    var valLen = _rsocketCore.UTF8Encoder.byteLength(val);
    var valBuf = (0, _rsocketCore.createBuffer)(valLen);
    _rsocketCore.UTF8Encoder.encode(val, valBuf, 0, valLen);

    var newEntries = aggregate.entries;
    newEntries.push({ keyLen: keyLen, keyBuf: keyBuf, valLen: valLen, valBuf: valBuf });

    return {
      //4 for the sizes plus the actual key and actual value
      totalSize: aggregate.totalSize + 4 + keyLen + valLen,
      entries: newEntries
    };
  }, { totalSize: 0, entries: [] });

  var offset = 0;
  var resultBuf = (0, _rsocketCore.createBuffer)(aggregatedTags.totalSize);
  aggregatedTags.entries.forEach(function (entry) {
    resultBuf.writeUInt16BE(entry.keyLen, offset);
    offset += 2; //2 bytes for key length

    _rsocketCore.BufferEncoder.encode(entry.keyBuf, resultBuf, offset, offset + entry.keyLen);
    offset += entry.keyLen;

    resultBuf.writeUInt16BE(entry.valLen, offset);
    offset += 2;

    _rsocketCore.BufferEncoder.encode(entry.valBuf, resultBuf, offset, offset + entry.valLen);
    offset += entry.valLen;
  });

  return resultBuf;
}

function bufferToMap(buffer) {
  var result = {};

  var offset = 0;
  while (offset < buffer.length) {
    var keyLen = buffer.readUInt16BE(offset);
    offset += 2;

    var key = _rsocketCore.UTF8Encoder.decode(buffer, offset, offset + keyLen);
    offset += keyLen;

    var valLen = buffer.readUInt16BE(offset);
    offset += 2;

    var value = _rsocketCore.UTF8Encoder.decode(buffer, offset, offset + valLen);
    offset += valLen;

    result[key] = value;
  }

  return result;
}

function trace(tracer, name, tags, references, startTime) {
  if (tracer && name) {
    return function (metadata) {
      (function (flowable) {
        flowable.lift(function (subscriber) {
          if (metadata) {
            if (!tags) {
              tags = {};
            }
            Object.keys(metadata).forEach(function (key) {
              tags[key] = map[key];
            });
          }
          return new _SpanSubscriber.SpanSubscriber(subscriber, tracer, name, null, tags, references, startTime);
        });
      });
    };
  } else {
    return function (map) {
      return function (publisher) {
        return publisher;
      };
    };
  }
}

function traceAsChild(tracer, name, tags, references, startTime) {
  if (tracer && name) {
    return function (context) {
      return function (flowable) {
        flowable.lift(function (subscriber) {
          return new _SpanSubscriber.SpanSubscriber(subscriber, tracer, name, context, tags, references, startTime);
        });
      };
    };
  } else {
    return function (context) {
      return function (publisher) {
        return publisher;
      };
    };
  }
}

function traceSingle(tracer, name, tags, references, startTime) {
  if (tracer && name) {
    return function (metadata) {
      (function (single) {
        if (metadata) {
          if (!tags) {
            tags = {};
          }
          Object.keys(metadata).forEach(function (key) {
            tags[key] = map[key];
          });
        }
        return new SpanSingle(single, tracer, name, null, tags, references, startTime);
      });
    };
  } else {
    return function (map) {
      return function (single) {
        return single;
      };
    };
  }
}

function traceSingleAsChild(tracer, name, tags, references, startTime) {
  if (tracer && name) {
    return function (context) {
      return function (flowable) {
        return new SpanSingle(single, tracer, name, null, tags, references, startTime);
      };
    };
  } else {
    return function (context) {
      return function (single) {
        return single;
      };
    };
  }
}
